proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,448 @@
+the compiler will function as if `c` was passed, and allocate the variables into the 
+global constant buffer.
+If the register type is 'c', it is allowed, but if `default_globals` is not set,
+`warn_hlsl_register_type_c_not_in_global_scope` will be emitted, and will be treated as an error by default.
+
+After this point, `default_globals` doesn't need to be set.
+If 't', 'u', or 's' are given when `basic` is set, then
+`err_hlsl_mismatching_register_type_and_variable_type` will be emitted.
+If any other register type is seen, `err_hlsl_unsupported_register_type_and_variable_type`
+will be emitted instead.
+
+Finally, in the case that `udt` is set, we first check `default_globals`. If it is set,
+then we can permit the 'c' register type if `contains_numeric` is set. If it is not set,
+`warn_hlsl_UDT_missing_basic_type` will be emitted, and treated as an error by default.
+After this point, `default_globals` doesn't need to be set.
+For every register type that is used to bind the resources contained in the given UDT, 
+we verify that the corresponding resource class flag has been set. These are the register types
+`t`, `u`, `b`, and `s`. If the corresponding resource class flag is not set,
+`warn_hlsl_UDT_missing_resource_type_member` will be emitted.
+Otherwise, if any other register type is given, then we emit `err_hlsl_unsupported_register_type_and_variable_type`.
+
+All the warnings introduced in this spec were not emitted in legacy versions of the compiler. The warnings,
+`warn_hlsl_register_type_c_not_in_global_scope`,
+`warn_hlsl_deprecated_register_type_b`,
+`warn_hlsl_deprecated_register_type_i`,
+`warn_hlsl_UDT_missing_resource_type_member`, and
+`warn_hlsl_UDT_missing_basic_type`, are all within a warning group known as 
+`disallow-legacy-binding-rules`. If legacy behavior is desired, a user can pass the `--Wno-disallow-legacy-binding-rules`
+flag to the compiler to silence the errors. When this flag is active, and legacy behavior is present, a warning will 
+be emitted instead of an error.
+
+Here is some pseudocode summarizing the diagnostic emission process:
+
+```
+if other is set:
+  emit err_hlsl_unsupported_register_type_and_variable_type
+
+if is_member is set:
+  emit err_hlsl_register_annotation_on_member
+
+if resource is set:
+  if register type does not match resource class flag:
+    emit err_hlsl_mismatching_register_type_and_resource_type
+
+if basic is set:
+  if default_globals is set:
+    if register type is 'b' or 'i':
+      emit warn_hlsl_deprecated_register_type_b or warn_hlsl_deprecated_register_type_i respectively
+
+  if register type is 'c'
+    if default_globals is not set:
+      emit warn_hlsl_register_type_c_not_in_global_scope
+  else if register type is 't' 'u' or 's'
+    emit err_hlsl_mismatching_register_type_and_variable_type
+  else emit err_hlsl_unsupported_register_type_and_variable_type
+
+if udt is set:

```

Comment:
I'm wondering if we can get rid of `udt` here and come up with something more general that just works on everything.

I also note that `basic` and `contains_numeric` mean _almost_ the same thing, and if we stopped differentiating between udt and non-udt then they would be the same thing.

Reply:
My initial reaction to this proposition is that it conflicts with the idea of not emitting any warning if a UDT doesn't have a resource type that's required by a specified register annotation. If UDT's are treated the same as standard resources, then we need to make some sort of distinction so that:
RWBuffer<int> r : register(s1); emits an error. Because if r's type was a UDT, and it didn't have a sampler, it wouldn't emit an error, but *because* it's not a UDT, it should emit an error.